### PR TITLE
Added the jest-junit reporter

### PR DESCRIPTION
**Pull request by wknapik**
Fixes https://github.com/brave/brave-browser/issues/19606
Matching devops PR https://github.com/brave/devops/pull/6289 (merged)
Security review https://github.com/brave/security/issues/665 (approved)

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: jest.config.js
Comment (petemill) [inline jest.config.js:82]: Do we need to selectively enable this only in CI via an env variable and/or exclude this file in .gitignore?

File: package.json
Comment (petemill) [inline package.json:351]: This should be in devDependencies

Comment (petemill) [reply package.json:351]: And needs a security review

**Review by petemill — CHANGES_REQUESTED** [review]

File: package.json
Comment (wknapik) [reply package.json:351]: Wow. I called
```
npm install --save --dev jest-junit
```
instead of
```
npm install --save-dev jest-junit
```
and npm, the great tool that it is, just ran with it. No warning, no error.

In fact, I see
```
npm install --hurr-durr jest-junit
```
also runs successfully. Smh.

**Comment by wknapik** [comment]
Seems some tests have an undefined test suite, which breaks linking in Jenkins reports (see: line 3):
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="jest tests" tests="10" failures="0" errors="0" time="10.711">
  <testsuite name="undefined" errors="0" failures="0" skipped="0" timestamp="2021-11-15T15:23:44" time="0.996" tests="3">
    <testcase classname=" generates a checksum" name=" generates a checksum" time="0.006">
    </testcase>
    <testcase classname=" checksum is stable" name=" checksum is stable" time="0.002">
    </testcase>
    <testcase classname=" checksum changes when file contents change" name=" checksum changes when file contents change" time="0.002">
    </testcase>
  </testsuite>
  <testsuite name="Apply Patches" errors="0" failures="0" skipped="0" timestamp="2021-11-15T15:23:44" time="3.517" tests="7">
    <testcase classname="Apply Patches applies simple patch to unmodified original" name="Apply Patches applies simple patch to unmodified original" time="0.283">
    </testcase>
    <testcase classname="Apply Patches does not apply patch to still-patched" name="Apply Patches does not apply patch to still-patched" time="0.32">
    </testcase>
    <testcase classname="Apply Patches resets target file if patch file is removed" name="Apply Patches resets target file if patch file is removed" time="0.337">
    </testcase>
    <testcase classname="Apply Patches handles missing file when resets target file if patch file is removed" name="Apply Patches handles missing file when resets target file if patch file is removed" time="0.452">
    </testcase>
    <testcase classname="Apply Patches handles bad patch file" name="Apply Patches handles bad patch file" time="0.353">
    </testcase>
    <testcase classname="Apply Patches handles no patch dir" name="Apply Patches handles no patch dir" time="0.272">
    </testcase>
    <testcase classname="Apply Patches handles no repo dir" name="Apply Patches handles no repo dir" time="0.304">
    </testcase>
  </testsuite>
</testsuites>
```

**Review by mihaiplesa — APPROVED** [review]

**Comment by wknapik** [comment]
Ready to merge, just needs approval.

File: build/commands/lib/calculateFileChecksum.test.js
Comment (petemill) [inline build/commands/lib/calculateFileChecksum.test.js:11]: Since this is a single-module test, lets call this `calculateFileChecksum`?

File: package-lock.json
Comment (petemill) [inline package-lock.json:21955]: This shouldn't be reverted back to the pre-security-fix state. Perhaps run the following:

```
git checkout master -- package-lock.json
npm install
```

That should fix this up.

**Review by petemill — CHANGES_REQUESTED** [review]
Also, please squash feedback commits in to the relative feature / fix commit

File: build/commands/lib/calculateFileChecksum.test.js
Comment (wknapik) [reply build/commands/lib/calculateFileChecksum.test.js]: Sure. Done.

File: package-lock.json
Comment (wknapik) [reply package-lock.json]: Done.

**Comment by wknapik** [comment]
@petemill what difference does the history on this branch make? We do squash+merge to the target branch in the PR anyway and the branch gets deleted.

File: package.json
Comment (wknapik) [reply package.json:318]: The security review is labeled `blocked/waiting-for-changes`, so this can't be merged yet anyway.
Once the audit findings are fixed in jest-junit, we'll just pin to the next version.

**Comment by petemill** [comment]
> @petemill what difference does the history on this branch make? We do squash+merge to the target branch in the PR anyway and the branch gets deleted.

@wknapik we do not do squash and merge in brave-core, and even if we did it wouldn't contain a good commit description. All these little commits absolutely do end up in the log for the repo and each file. Having good minimal 1 commit-per-feature merged to master ensures:
- The commit log of a file tells a journey of the file / repo and not a journey of your PR e.g. `Enable jest to report test results to CI server` vs `Added missing change.`
- We don't permanently store unnecessary file changes inside everyone's .git directory

**Comment by wknapik** [comment]
@petemill 

> we do not do squash and merge in brave-core, and even if we did it wouldn't contain a good commit description.

When you click `Squash and merge` in a GithHub PR, you get to fill in a short and long message. The default for the short one is the PR title and the default for the long one is a list of all commit messages on that branch as bullet points. But you get to edit those before actually merging.

Also, as I'm typing this, the button above says "Squash and merge".

> Having good minimal 1 commit-per-feature merged to master ensures:

Sure, the benefits are clear. And they are achieved if you use GitHub's squash+merge. We use it in the devops repo, works great.

The problem with not using the GH feature is that you either keep squashing as you develop (definitely don't like that), or you risk forgetting to do it when the time comes to merge (even worse).

Anyway, I'll squash commits on this branch in a minute.

File: package.json
Comment (mihaiplesa) [reply package.json:318]: that doesn't seem to be coming soon so let's revert this to be greater than till they release a new version

Comment (wknapik) [reply package.json]: Done.

**Comment by wknapik** [comment]
[Approved for merging](https://github.com/brave/security/issues/665#issuecomment-973148015) despite the outstanding finding in jest-junit.

Please take another look at package-lock.json, which contains more than I'd expect. If it's ok, we're good to merge.

**Comment by wknapik** [comment]
Can I get an approval from the dev side? Wanna make sure the package-lock.json changes are ok before merging.

**Review by petemill — APPROVED** [review]
Thanks for also cleaning up 👍

**Pull request merged**
